firedrake/interpolation.py

```diff
@@ -1664,3 +1666,120 @@ def multHermitian(self, source_vec, target_vec):
             # matrix will then have rows of zeros for those points.
             target_vec.zeroEntries()
             self.reduce(source_vec, target_vec)
+
+
+class ClementInterpolator(SameMeshInterpolator):
+    r"""
+    Compute the Clément interpolant of a :math:`\mathbb{P}0` source field, i.e., take
+    the volume average over neighbouring cells at each vertex.
+
+    See :cite:`Clement1975` for details.
+
+    For arguments, see :class:`.Interpolator`.
+    """
+
+    def __new__(cls, expr, V, **kwargs):
+        target_mesh = as_domain(V)
+        source_mesh = extract_unique_domain(expr) or target_mesh
+        if target_mesh is source_mesh:
+            return object.__new__(ClementInterpolator)
+        else:
+            raise ValueError("Clément interpolation requires the source and target meshes to coincide.")
+
+    @no_annotations
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_dofs=False,
+    ):
+        if subset:
+            raise NotImplementedError("subset not implemented")
+        if freeze_expr:
+            raise NotImplementedError("freeze_expr not implemented")
+        if access != op2.WRITE:
+            raise NotImplementedError("access other than op2.WRITE not implemented")
+        if bcs:
+            raise NotImplementedError("bcs not implemented")
+        element = V.ufl_element()
+        if element.family() != "Lagrange" or element.degree() != 1:
+            raise ValueError("Clément interpolation must target a P1 space.")
+        super().__init__(expr, V, subset=subset, freeze_expr=freeze_expr, access=access,
+                         bcs=bcs, allow_missing_dofs=allow_missing_dofs)
+
+    @PETSc.Log.EventDecorator()
+    def interpolate(self, function, output=None, adjoint=False):
+        """
+        Compute the Clément interpolant applied to a source function.
+
+        Parameters
+        ----------
+        function: firedrake.function.Function or firedrake.cofunction.Cofunction
+                  Function to be interpolated.
+        output: firedrake.function.Function or firedrake.cofunction.Cofunction
+                A function to contain the output.
+        adjoint: bool
+                   Set to true to apply the adjoint of the interpolation
+                   operator.
+
+        Returns
+        -------
+        firedrake.function.Function or firedrake.cofunction.Cofunction
+            The resulting interpolated function.
+        """
+        from firedrake.assemble import assemble
```

Comment:
```suggestion
```

Reply:
Done in https://github.com/firedrakeproject/firedrake/pull/4244/commits/6e67568a7eb9c2aaf39864c1d586d88caffe42e6.